drivers/bluetooth/hci/ipm_stm32wb.c

```diff
@@ -0,0 +1,398 @@
+
+#define POOL_SIZE (CFG_TLBLE_EVT_QUEUE_LENGTH * 4 * \
+		DIVC((sizeof(TL_PacketHeader_t) + TL_BLE_EVENT_FRAME_SIZE), 4))
+
+/* Private variables ---------------------------------------------------------*/
+PLACE_IN_SECTION("MB_MEM1") ALIGN(4) static TL_CmdPacket_t BleCmdBuffer;
+PLACE_IN_SECTION("MB_MEM2") ALIGN(4) static u8_t EvtPool[POOL_SIZE];
+PLACE_IN_SECTION("MB_MEM2") ALIGN(4) static TL_CmdPacket_t SystemCmdBuffer;
+PLACE_IN_SECTION("MB_MEM2") ALIGN(4) static u8_t
+	SystemSpareEvtBuffer[sizeof(TL_PacketHeader_t) + TL_EVT_HDR_SIZE + 255];
+PLACE_IN_SECTION("MB_MEM2") ALIGN(4) static u8_t
+	BleSpareEvtBuffer[sizeof(TL_PacketHeader_t) + TL_EVT_HDR_SIZE + 255];
+PLACE_IN_SECTION("MB_MEM2") ALIGN(4) static u8_t
+	HciAclDataBuffer[sizeof(TL_PacketHeader_t) + 5 + 251];
+
+static void syscmd_status_not(SHCI_TL_CmdStatus_t status);
+static void sysevt_received(void *pdata);
+
+#define BT_DBG_ENABLED IS_ENABLED(CONFIG_BT_DEBUG_HCI_DRIVER)
+#define LOG_MODULE_NAME hci_ipm
+#include "common/log.h"
+
+#define HCI_CMD                 0x01
+#define HCI_ACL                 0x02
+#define HCI_SCO                 0x03
+#define HCI_EVT                 0x04
+
+static K_SEM_DEFINE(c2_started, 0, 1);
+static K_SEM_DEFINE(ble_sys_wait_cmd_rsp, 0, 1);
+static K_SEM_DEFINE(acl_data_ack, 1, 1);
+static K_SEM_DEFINE(ipm_busy, 1, 1);
+
+struct aci_set_tx_power {
+	u8_t cmd;
+	u8_t value[2];
+};
+
+#define ACI_WRITE_SET_TX_POWER_LEVEL       BT_OP(BT_OGF_VS, 0xFC0F)
+
+static void stm32wb_start_ble(void)
+{
+	SHCI_C2_Ble_Init_Cmd_Packet_t ble_init_cmd_packet =
+	{ { { 0, 0, 0 } },                     /**< Header unused */
+	  { 0,                                 /** pBleBufferAddress not used */
+	    0,                                 /** BleBufferSize not used */
+	    CFG_BLE_NUM_GATT_ATTRIBUTES,
+	    CFG_BLE_NUM_GATT_SERVICES,
+	    CFG_BLE_ATT_VALUE_ARRAY_SIZE,
+	    CFG_BLE_NUM_LINK,
+	    CFG_BLE_DATA_LENGTH_EXTENSION,
+	    CFG_BLE_PREPARE_WRITE_LIST_SIZE,
+	    CFG_BLE_MBLOCK_COUNT,
+	    CFG_BLE_MAX_ATT_MTU,
+	    CFG_BLE_SLAVE_SCA,
+	    CFG_BLE_MASTER_SCA,
+	    CFG_BLE_LSE_SOURCE,
+	    CFG_BLE_MAX_CONN_EVENT_LENGTH,
+	    CFG_BLE_HSE_STARTUP_TIME,
+	    CFG_BLE_VITERBI_MODE,
+	    CFG_BLE_LL_ONLY,
+	    0 } };
+
+	/**
+	 * Starts the BLE Stack on CPU2
+	 */
+	SHCI_C2_BLE_Init(&ble_init_cmd_packet);
+}
+
+static void sysevt_received(void *pdata)
+{
+	k_sem_give(&c2_started);
+}
+
+static void syscmd_status_not(SHCI_TL_CmdStatus_t status)
+{
+	BT_DBG("status:%d", status);
+}
+
+void TM_EvtReceivedCb(TL_EvtPacket_t *hcievt)
+{
+	struct net_buf *buf;
+	struct bt_hci_acl_hdr acl_hdr;
+	TL_AclDataSerial_t *acl;
+
+	k_sem_take(&ipm_busy, K_NO_WAIT);
+
+	switch (hcievt->evtserial.type) {
+	case HCI_EVT:
+		BT_DBG("EVT: hcievt->evtserial.evt.evtcode: %d",
+		       hcievt->evtserial.evt.evtcode);
+		switch (hcievt->evtserial.evt.evtcode) {
+		case BT_HCI_EVT_VENDOR:
+			/* Vendor events are currently unsupported */
+			BT_ERR("Unknown evtcode type %d",
```

Comment:
suggestion, use `BT_ERR("Unknown evtcode type 0x%02x"` to have it equal to hci_core